File: apps/ios/Velix/Core/APIClient.swift
import Foundation

enum APIError: Error, LocalizedError {
    case http(status: Int, message: String?, reason: String?)
    case decoding(Error)
    case network(Error)

    var errorDescription: String? {
        switch self {
        case .http(_, let message, _): return message ?? "Erro ao falar com o servidor"
        case .decoding: return "Resposta inesperada do servidor"
        case .network: return "Sem conexão com o servidor"
        }
    }
}

final class APIClient {
    let baseURL: URL
    var token: String?

    init(baseURL: URL, token: String? = nil) {
        self.baseURL = baseURL
        self.token = token
    }

    func makeRequest(path: String, method: String) -> URLRequest {
        // O backend expõe tudo sob /api (ver `app.setGlobalPrefix('api')` em
        // apps/api/src/main.ts) — sem isso toda chamada cai no Next.js do
        // painel web e volta 404, inclusive o próprio /auth/login.
        var request = URLRequest(url: URL(string: "/api" + path, relativeTo: baseURL)!)
        request.httpMethod = method
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return request
    }

    private func send<T: Decodable>(_ request: URLRequest) async throws -> T {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw APIError.network(error)
        }
        guard let http = response as? HTTPURLResponse else {
            throw APIError.network(URLError(.badServerResponse))
        }
        guard (200..<300).contains(http.statusCode) else {
            let body = try? JSONDecoder().decode(APIErrorBody.self, from: data)
            throw APIError.http(status: http.statusCode, message: body?.message, reason: body?.reason)
        }
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw APIError.decoding(error)
        }
    }

    func get<T: Decodable>(_ path: String) async throws -> T {
        try await send(makeRequest(path: path, method: "GET"))
    }

    func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = makeRequest(path: path, method: "POST")
        request.httpBody = try JSONEncoder().encode(body)
        return try await send(request)
    }

    /// Ações sem corpo (start/stop/restart) — sem isso cada chamador teria
    /// que inventar seu próprio `struct Empty: Encodable {}`.
    func post<T: Decodable>(_ path: String) async throws -> T {
        try await send(makeRequest(path: path, method: "POST"))
    }

    func put<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = makeRequest(path: path, method: "PUT")
        request.httpBody = try JSONEncoder().encode(body)
        return try await send(request)
    }

    func patch<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = makeRequest(path: path, method: "PATCH")
        request.httpBody = try JSONEncoder().encode(body)
        return try await send(request)
    }

    func delete(_ path: String) async throws {
        let _: EmptyResponse = try await send(makeRequest(path: path, method: "DELETE"))
    }
}

struct EmptyResponse: Decodable {}

File: apps/android/app/src/main/java/com/velix/app/core/ApiClient.kt
package com.velix.app.core

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json

sealed class ApiException(message: String) : Exception(message) {
    class Http(val status: Int, val serverMessage: String?, val reason: String?) :
        ApiException(serverMessage ?: "Erro ao falar com o servidor")
    class Network(cause: Throwable) : ApiException(cause.message ?: "Sem conexão com o servidor")
    class Decoding(cause: Throwable) : ApiException("Resposta inesperada do servidor")
}

class ApiClient(
    @PublishedApi internal val baseUrl: String,
    @PublishedApi internal var token: String? = null,
) {
    @PublishedApi
    internal val json = Json { ignoreUnknownKeys = true }

    @PublishedApi
    internal val client = HttpClient(OkHttp) {
        install(ContentNegotiation) { json(json) }
    }

    @PublishedApi
    internal suspend inline fun <reified T> request(
        path: String,
        method: HttpMethod,
        body: Any? = null,
    ): T {
        val response = try {
            client.request("$baseUrl$path") {
                this.method = method
                token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
                if (body != null) {
                    contentType(ContentType.Application.Json)
                    setBody(body)
                }
            }
        } catch (e: Exception) {
            throw ApiException.Network(e)
        }

        if (!response.status.isSuccess()) {
            val errorBody = try {
                json.decodeFromString<ApiErrorBody>(response.bodyAsText())
            } catch (e: Exception) {
                null
            }
            throw ApiException.Http(response.status.value, errorBody?.message, errorBody?.reason)
        }

        return try {
            response.body()
        } catch (e: Exception) {
            throw ApiException.Decoding(e)
        }
    }

    suspend inline fun <reified T> get(path: String): T = request(path, HttpMethod.Get)
    suspend inline fun <reified T> post(path: String, body: Any): T = request(path, HttpMethod.Post, body)
    suspend inline fun <reified T> put(path: String, body: Any): T = request(path, HttpMethod.Put, body)
    suspend fun delete(path: String) { request<Unit>(path, HttpMethod.Delete) }
}
